docs/sql/types.md

```diff
@@ -23,6 +23,7 @@ The compiler supports the following SQL data types:
 | `TIMESTAMP`                 | A value containing a date and a time without a timezone.                                                                                                           | `DATETIME`                 |
 | `DATE`                      | A date value.                                                                                                                                                      |                            |
 | `GEOMETRY`                  | A geographic data type (only rudimentary support at this point).                                                                                                   |                            |
+| `ROW`                       | A tuple (anonymous struct with named fields) with 1 or more elements.  Example `ROW(left int null, right varchar)`                                                 |                            |
```

Comment:
Can `ROW` types appear in table declarations? Can they be generated as part of a view? If so, they may not work with all of our adapters.

Reply:
In principle, yes, they can, and the can be also fields of structs, array elements, or map keys.
You see why adding a new type is not easy.

Reply:
Then we shouldn't be adding this feature without tests. Would it be possible to disallow that?

Reply:
But these aren't that different from the existing structs and tuples (except they can now be nested to any depth).
So maybe the adapters will work.

Reply:
As you can see I added at least a dozen tests.
I can mark this under construction until you have more tests.

Reply:
It's easy for the compiler to reject ROW in input and output tables for now, if that's what you mean.

Reply:
I meant specifically input/output tests. 

> It's easy for the compiler to reject ROW in input and output tables for now, if that's what you mean.

yes

Reply:
I have pushed a commit which does this.
There is a compiler flag to disable the restriction, which can enable testing for this feature - assuming we build a solution for #2898 first

Reply:
Turns out I cannot do this, I cannot distinguish between user-defined types and ROW types, they are compiled into the same objects. Since UDTs work, let's allow ROW types too.